### PR TITLE
refactor node state to fully put the agent in charge

### PR DESCRIPTION
**Pull request by bmc-msft**
This moves the node state management in the hands of the agent, and we tell the agent when to shut down, etc.

### CODE REVIEW

File: src/api-service/__app__/onefuzzlib/pools.py [hunk 1]
```diff
@@ -154,16 +160,87 @@ def stop_task(cls, task_id: UUID) -> None:
...
+        )
+        stop_message.save()
+
+    def to_reimage(self, done: bool = False) -> None:
+        if done:
+            if self.state not in NodeState.ready_for_reset():
+                self.state = NodeState.done
+
+        if not self.reimage_requested and not self.delete_requested:
+            self.reimage_requested = True
```
Comment (chkeita) [inline src/api-service/__app__/onefuzzlib/pools.py:227]: Is this reset after the reimage ?

Comment (bmc-msft) [reply src/api-service/__app__/onefuzzlib/pools.py:227]: It is now that #86 has been merged, I was able to enable it.

File: src/api-service/__app__/agent_events/__init__.py [hunk 2]
```diff
@@ -47,11 +47,22 @@ def get_node_checked(machine_id: UUID) -> Node:
...
+        logging.info("stopping free node with reset flags: %s", node.machine_id)
+        node.stop()
+        return
+
+    if state == NodeState.init:
+        if node.delete_requested:
+            node.stop()
+            return
+        node.reimage_requested = False
+        node.save()
```
Comment (chkeita) [inline src/api-service/__app__/agent_events/__init__.py:64]: I think this logic should be in the node class.

File: src/pytypes/onefuzztypes/models.py [hunk 3]
```diff
@@ -399,6 +399,8 @@ class Node(BaseModel):
     scaleset_id: Optional[UUID] = None
     tasks: Optional[List[Tuple[UUID, NodeTaskState]]] = None
     version: str = Field(default="1.0.0")
+    reimage_requested: bool = Field(default=False)
+    delete_requested: bool = Field(default=False)
```
Comment (chkeita) [inline src/pytypes/onefuzztypes/models.py:403]: could both of these have been part of the node states

**Review by chkeita — APPROVED** [review]

File: src/pytypes/onefuzztypes/models.py (hunk 3 above)
Comment (bmc-msft) [reply src/pytypes/onefuzztypes/models.py:403]: We did that previously, but then got into issues where we were out of sync with the supervisor.  This treats the supervisor as having ground truth knowledge of node.state, but still lets us remember when a node should be reimaged/deleted.

File: src/api-service/__app__/agent_events/__init__.py (hunk 2 above)
Comment (bmc-msft) [reply src/api-service/__app__/agent_events/__init__.py:64]: probably, but I will do that in another PR

File: src/api-service/__app__/onefuzzlib/pools.py [hunk 4]
```diff
@@ -99,7 +105,7 @@ def search_outdated(
         # azure table query always return false when the column does not exist
         # We write the query this way to allow us to get the nodes where the
         # version is not defined as well as the nodes with a mismatched version
-        version_query = "not (version ne '%s')" % __version__
+        version_query = "not (version eq '%s')" % __version__
```
Comment (chkeita) [inline src/api-service/__app__/onefuzzlib/pools.py:108]: 👍

**Pull request merged**